### PR TITLE
Paragraph on API dynamics to address issue #5

### PR DESCRIPTION
**Pull request by mirjak**
I didn't add a new section because there are a few hints on dynamic all over the doc already and especially here there was already text that talked about initiate and listen. So I thought that was a good fit here. However, if you want to rather have this in a separate section we have to probably go though the whole doc again and see what we already have said somewhere....

### CODE REVIEW

File: draft-trammell-taps-post-sockets.md [hunk 1]
```diff
@@ -585,6 +585,24 @@ simplified interface which uses default message properties (reliable transport
...
+For each connection between a Local and a Remote a new Carrier is created and 
+destroyed when the connection is closed. However, a new Carrier may use an existing
+Association if present for the requested Local-Remote pair and permitted by the
+PolicyContext that can be provided at Carrier initiation. Further the system-wide
+PolicyContext can contain more information that determine when to create or destroy
+Associations other than at Carrier initiation. E.g. an Association can be created 
+at system start, based on the configured PolicyContext or also by a manual action of 
+an single application, for Local-Remote pairs that are known to be likely used soon, and 
+to pre-establish, e.g., cryptographic context as well as potentially collect current information
+about path capabilities. Every time an actually connection with a specific PSI is established between the
```
Comment (tfpauly) [inline draft-trammell-taps-post-sockets.md]: actually -> actual?

File: draft-trammell-taps-post-sockets.md [hunk 2]
```diff
@@ -585,6 +585,24 @@ simplified interface which uses default message properties (reliable transport
...
+PolicyContext can contain more information that determine when to create or destroy
+Associations other than at Carrier initiation. E.g. an Association can be created 
+at system start, based on the configured PolicyContext or also by a manual action of 
+an single application, for Local-Remote pairs that are known to be likely used soon, and 
+to pre-establish, e.g., cryptographic context as well as potentially collect current information
+about path capabilities. Every time an actually connection with a specific PSI is established between the 
+Local and Remote, the Association learns new Path information and stores them. This
+information can be used when a new transient is created, e.g. to decide which PSI
+to use (to provide the highest probably for a successful connection attempt) or which
+PSIs to probe for (first). A Transient is created when an application actually sends a
```
Comment (tfpauly) [inline draft-trammell-taps-post-sockets.md]: A transient -> One or more transients

Comment (tfpauly) [reply draft-trammell-taps-post-sockets.md]: Ah you mention that there can multiple in the next sentence...

**Comment by mirjak** [comment]
Wouldn't it make maybe more sense to show these steps based on examples? Isn't this what we already have for the simple case (without idempotent data) in the first example?

**Comment by tfpauly** [comment]
That's a good point—this does end up overlapping a lot with the example section. One difference is that the example is in some form of code block, as opposed to a prose/list description of the state machine and steps.

Perhaps this is a bigger question—should the heart of the examples be in code, or be a walkthrough of the possible state machine steps and transitions? Which API calls are valid at any given point? Not just an example, but the official description of which dynamics are allowed to happen or not.

**Comment by mirjak** [comment]
So to your last question about which api call are valid at which point. I think that can actually as best be discussed with code, however maybe we should not see the code only as example code but as something to support the spec? I think that would maybe mean renaming the example section and add some more explanations...?

**Comment by tfpauly** [comment]
OK, I see that code is a good way to explain. However, for the full "dynamics" we may need to explore the error cases, and all the possible things to call within the examples. Things like which functions are permitted to be called when in relation to one another, what possible states can be returned, etc.

**Comment by britram** [comment]
this also seems mergeable to me in its current state. for errors etc -- maybe what we want is a control sequence diagram? I mean, if we didn't need to do ASCII art I'd say it's *definitely* what we want.

**Pull request merged**